packages/kbn-server-route-repository/src/register_routes.test.ts

```diff
@@ -167,38 +177,226 @@ describe('registerRoutes', () => {
       status: 200,
-      payload: 'internal',
-      options: { body: 'internal' },
+      payload: 'result',
+      options: { body: 'result' },
     });
   });
 
   it('allows for route handlers to define a custom response', async () => {
-    const [_, publicRouteHandler] = getAddVersion.mock.calls[0];
-    const publicResult = await publicRouteHandler(mockContext, mockRequest, kibanaResponseFactory);
+    const handler = jest
+      .fn()
+      .mockResolvedValue(
+        kibanaResponseFactory.custom({ statusCode: 201, body: { message: 'result' } })
+      );
+
+    callRegisterRoutes({
+      'POST /internal/route': {
+        endpoint: 'POST /internal/route',
+        handler,
+      },
+    });
 
-    expect(publicHandler).toHaveBeenCalledTimes(1);
-    expect(publicResult).toEqual({ status: 201, payload: { message: 'public' }, options: {} });
+    const [_, wrappedHandler] = post.mock.calls[0];
+    const result = await wrappedHandler(mockContext, mockRequest, kibanaResponseFactory);
+
+    expect(handler).toHaveBeenCalledTimes(1);
+    expect(result).toEqual({ status: 201, payload: { message: 'result' }, options: {} });
   });
 
   it('translates errors thrown in a route handler to an error response', async () => {
-    const [_, errorRouteHandler] = get.mock.calls[1];
-    const errorResult = await errorRouteHandler(mockContext, mockRequest, kibanaResponseFactory);
+    const handler = jest.fn().mockRejectedValue(new Error('error'));
+
+    callRegisterRoutes({
+      'POST /internal/route': {
+        endpoint: 'POST /internal/route',
+        handler,
+      },
+    });
 
-    expect(errorHandler).toHaveBeenCalledTimes(1);
-    expect(errorResult).toEqual({
+    const [_, wrappedHandler] = post.mock.calls[0];
+    const error = await wrappedHandler(mockContext, mockRequest, kibanaResponseFactory);
+
+    expect(handler).toHaveBeenCalledTimes(1);
+    expect(error).toEqual({
       status: 500,
       payload: { message: 'error', attributes: { data: {} } },
       options: {},
     });
   });
+
+  describe('when using zod', () => {
+    const makeZodValidationObjectSpy = jest.spyOn(
+      makeZodValidationObject,
+      'makeZodValidationObject'
+    );
+
+    const zodParamsRt = z.object({
+      body: z.object({
+        bodyParam: z.string(),
+      }),
+      query: z.object({
+        queryParam: z.string(),
+      }),
+      path: z.object({
+        pathParam: z.string(),
+      }),
+    });
+
+    it('uses Core validation', () => {
+      callRegisterRoutes({
+        'POST /internal/route': {
+          endpoint: 'POST /internal/route',
+          params: zodParamsRt,
+          handler: jest.fn,
+        },
+      });
+
+      const [internalRoute] = post.mock.calls[0];
+      expect(makeZodValidationObjectSpy).toHaveBeenCalledWith(zodParamsRt);
+      expect(internalRoute.validate).toEqual(makeZodValidationObjectSpy.mock.results[0].value);
+    });
+
+    it('passes on params', async () => {
```

Comment:
Again, testing data flow because the core part is mocked so I cannot test that it actually uses the schema for validation. 

In `makeZodValidationObject` I'm checking that it's turning it into strict, so adding a new test for that feels like testing zod rather than our code.